### ISSUE TITLE
(v2) LoadClientStateMiddleware passes through a nil request

### ISSUE THREAD

**Issue by natefinch** [volatiletech/authboss#181]
LoadClientStateMiddleware calls LoadClientState.
LoadClientState is returning a nil http.Request if either session state or cookie state is nil.   IMO, this function should never return a nil request.. that's almost certain to cause a nil pointer panic in further handlers.  This is what is happening to me.

https://github.com/volatiletech/authboss/blob/v2/client_state.go#L126
 
I'm not sure what the correct fix is... but returning a nil request without it being an error is definitely wrong here.

**Comment by aarondl** [comment]
Yeah. That's seems totally insane. The way it should work is that if the state is not nil, it should modify `r` and allow the code to fall down to the next if statement to modify `r` more, the same thing happens there, and finally the original or modified `r`, is always returned at the end.

Having fixed this: https://github.com/volatiletech/authboss-clientstate/issues/2

I've found a new issue related in the same area. I believe the dropped error was intentional now, and that if session loading is failing for whatever reason (invalid session, expired timestamp (the error I'm getting atm), users can produce crazy inputs) we just ignore that error and return nil for the state because we don't care why we couldn't load it 99% of the time.

To log these errors seems perhaps overzealous. So the only thing to do is to revert the fix to client-state/2, and the API looks like: "If you can't load the session, it's not an error unless you couldn't load it for a show-stopper serious reason which means you'll never be able to load sessions again", or to log every session loading error that's passed back instead of panic'ing.

Currently the route I'm looking at is logging. Is it truly an error level log though? Tough calls.

**Comment by aarondl** [comment]
It's possible that the proper fix for client-state should have been to switch on `IsUsage` and `IsDecode` errors, ignoring `IsDecode`.

http://www.gorillatoolkit.org/pkg/securecookie#Error

**Comment by natefinch** [comment]
yeah, so this was originally hit because I forgot to set a cookie store key and a session store key... so they were being called with an empty []byte .... which evidently makes the clientstate return nil....

**Comment by aarondl** [comment]
So I've fixed the issue that you've mentioned here in the exact way that I mentioned (because it was insane how it was happening before).

I've also touched up the error handling in the `clientstate` package to only return errors that are caused by non-decoding problems. That way your key missing error should still log problems and die on requests, but if John has a remember me cookie from a decade ago, it'll just be ignored.

Authboss rc6 has been tagged with the changes. Although as you know clientstate is a bit more of a wild-west repo, so you'll just have to fetch master of that :) Thanks for reporting all these issues.
